SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp

```diff
@@ -0,0 +1,182 @@
+	return bIsPingEnabled;
+}
+
+void USpatialPingComponent::SetPingEnabled(bool bSetEnabled)
+{
+	// Only execute if this component is attached to a player controller.
+	if (OwningController != nullptr)
+	{
+		// Only execute on owning local client.
+		if (OwningController->GetLocalRole() == ENetRole::ROLE_AutonomousProxy)
+		{
+			if (bSetEnabled && !bIsPingEnabled)
+			{
+				EnablePing();
+			}
+			else if (!bSetEnabled && bIsPingEnabled)
+			{
+				DisablePing();
+			}
+		}
+	}
+}
+
+float USpatialPingComponent::GetPing() const
+{
+	return RoundTripPing;
+}
+
+void USpatialPingComponent::EnablePing()
+{
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		LastSentPingID = 0;
+		TimeoutCount = 0;
+		// Set looping timer to 'tick' this component with frequency matching minimum ping interval.
+		World->GetTimerManager().SetTimer(PingTickHandle, this, &USpatialPingComponent::TickPingComponent, MinPingInterval, true);
+		bIsPingEnabled = true;
+	}
+}
+
+void USpatialPingComponent::DisablePing()
+{
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		bIsPingEnabled = false;
+		// Clear the timers.
+		World->GetTimerManager().ClearTimer(PingTickHandle);
+		World->GetTimerManager().ClearTimer(PingTimerHandle);
+		// Reset ping output value.
+		RoundTripPing = 0.f;
+	}
+}
+
+void USpatialPingComponent::TickPingComponent()
+{
+	SendNewPing();
+	// Pass latest measured ping to owning controller to be processed by PlayerState.
+	OwningController->UpdatePing(RoundTripPing);
+}
+
+void USpatialPingComponent::SendNewPing()
+{
+	// Generate a new ping ID.
+	uint16 NewPingID = LastSentPingID + 1;
+	// Send and record new ping ID.
+	SendServerWorkerPingID(NewPingID);
+	LastSentPingID = NewPingID;
+	LastSentPingTimestamp = FPlatformTime::Seconds();
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		// Set a timeout timer to await a reply.
+		World->GetTimerManager().SetTimer(PingTimerHandle, this, &USpatialPingComponent::OnPingTimeout, TimeoutLimit, false);
+	}
+}
+
+void USpatialPingComponent::OnPingTimeout()
+{
+	UWorld* World = GetWorld();
+	if (World != nullptr)
+	{
+		// Clear the timeout timer.
+		World->GetTimerManager().ClearTimer(PingTimerHandle);
+	}
+	// Update round trip ping to reflect timeout.
+	TimeoutCount++;
+	RoundTripPing = TimeoutLimit * TimeoutCount;
+	// Attempt to send another ping.
+	SendNewPing();
+}
+
+void USpatialPingComponent::OnRep_ReplicatedPingID()
+{
+	// Check that replicated ping ID matches the last sent AND check ping is enabled to catch cases where a value is replicated after being locally disabled.
+	if (ReplicatedPingID == LastSentPingID && bIsPingEnabled)
+	{
+		UWorld* World = GetWorld();
+		if (World != nullptr)
+		{
+			// Clear the timeout timer.
+			World->GetTimerManager().ClearTimer(PingTimerHandle);
+			TimeoutCount = 0;
+		}
+		// Calculate the round trip ping
+		RoundTripPing = static_cast<float>(FPlatformTime::Seconds() - LastSentPingTimestamp);
+		if (RoundTripPing >= MinPingInterval)
```

Comment:
I'm not sure if we can end up in this situation. If `MinPingInterval` passed since last ping, we would have already sent a new ping and `LastSentPingID` would have increased, so the if condition `ReplicatedPingID == LastSentPingID` would fail.

Reply:
This conditional catches a response that falls between the minimum interval and the timeout and should be required now that I'm not erroneously calling the SendNewPing each tick.